scripts/version_buckets.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+"""
+Requires boto3 and click packages to be installed.
+
+1) Create a new virtualenv (python3 -m venv my_venv).
+2) Enable the virtualenv (my_venv/bin/activate).
+3) Install the dependencies: (pip install boto3 click).
+4) Run the command (--help will give you more information).
+
+What does this script do?
+
+Reads all the current buckets, ensures that versioning is switched on for those
+that don't yet have it and sets the life cycle configuration to send
+non-current versions of files to glacier storage after 30 days.
+
+This script should only be run once, when the following PR lands in master in
+the control panel:
+
+https://github.com/ministryofjustice/analytics-platform-control-panel/pull/816
+
+(From that moment on, all S3 buckets will be configured in this way by the
+control panel, as a default.)
+
+This script allows us to preview the potential changes and run the script on
+a defined number of buckets at once.
+"""
+import os
+import click
+import logging
+import sys
+import boto3
+import json
+from botocore.exceptions import ClientError
+from pathlib import Path
+
+
+# A flag to indicate if the command is to be run in verbose mode.
+VERBOSE = False
+
+
+# Setup logging. Logs will appear in user's home directory.
+LOGFILE = os.path.join(str(Path.home()), "bucket_update.log")
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logfile_handler = logging.FileHandler(LOGFILE)
+log_formatter = logging.Formatter("%(levelname)s: %(message)s")
+logfile_handler.setFormatter(log_formatter)
+logger.addHandler(logfile_handler)
+
+
+def is_control_panel_bucket(name, bucket):
+    """
+    Returns a boolean indication to say if the referenced name/bucket is
+    associated with the control panel app.
+
+    They either start with the name "alpha-app-" or they have a tag of
+    "buckettype" with the value "datawarehouse".
+
+    See discussion here:
+
+    https://github.com/ministryofjustice/analytics-platform-ops/pull/320#discussion_r433765261
+    """
+    is_valid = name.startswith("alpha-app-")
+    if name.startswith("dev-"):
+        # Ignore this since it's a bucket associated with the dev instance.
+        is_valid = False
+    elif not is_valid:
```

Comment:
I think there's unnecessary cognitive load in reading lines like this. What does 'valid' mean in this context? I get the impression you're working too hard to have a single return statement, either because there is a rule of thumb to achieve that, or because the log statement.
```
if name.startswith("alpha-app-"):
    # these are all the App Data Sources
    return True
if name.startswith("dev-"):
    # not interested in that environment
    return False
```
I think the 'single return statement' rule is not hard and fast:
https://stackoverflow.com/questions/36707/should-a-function-have-only-one-return-statement
and you could move the log statement to move to where you call this function.

Reply:
I tend to agree with that.
I usually prefer to deal with the exceptional/trivial cases early on and leave the rest of the function for the main logic. It makes things more readable, e.g.

```python
if not name.startswith("alpha-"):
    # not interested in other environments
    return False

if name.startswith("alpha-app-"):
    # these are all the App Data Sources
    return True

# Rest of the logic which check `buckettype=datawarehouse` tag
# ...
```

I find this easier to read as I don't have to track down where/if the `is_valid` variable changed but I'm a bit slow so maybe it's just me 😋.

Reply:
In this case, the heuristics are not as simple as yes/no -- there's an "and" in the check in there somewhere (hence the need of the flag). Also, it seemed prudent to log the value of the flag so we had a record somewhere. The code does this at the end of this function before returning the result.